gno.land/cmd/gnoland/main.go

```diff
@@ -314,3 +309,103 @@ func loadGenesisBalances(path string) []string {
 	}
 	return balances
 }
+
+// listGnoPkgs lists all gno packages in the given root directory.
+func listGnoPkgs(root string) ([]pkg, error) {
+	var pkgs []pkg
+
+	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() {
+			goModPath := filepath.Join(path, "gno.mod")
+			data, err := os.ReadFile(goModPath)
+			if err == nil {
+				gnoMod, err := gnomod.Parse(goModPath, data)
+				if err != nil {
+					return fmt.Errorf("parse: %w", err)
+				}
+				gnoMod.Sanitize()
+				if err := gnoMod.Validate(); err != nil {
+					return fmt.Errorf("validate: %w", err)
+				}
+				pkgs = append(pkgs, pkg{
+					name: gnoMod.Module.Mod.Path,
+					path: path,
+					requires: func() []string {
+						var reqs []string
+						for _, req := range gnoMod.Require {
+							reqs = append(reqs, req.Mod.Path)
+						}
+						return reqs
+					}(),
+				})
+				return fs.SkipDir
+			} else if !os.IsNotExist(err) {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pkgs, nil
+}
+
+// sortPkgs sorts the given packages by their dependencies.
+func sortPkgs(pkgs []pkg) error {
+	visited := make(map[string]bool)
+	onStack := make(map[string]bool)
+	sortedPkgs := make([]pkg, 0, len(pkgs))
+
+	var visit func(pkg pkg) error
+	visit = func(pkg pkg) error {
```

Comment:
```suggestion
	visit := func(pkg pkg) error {
```

Reply:
we are calling `visit()` recursively. can't short assign it.

Reply:
then, can you extract is as a named function for clarity?

Reply:
Can we keep it current way? Since it's not going to be used anywhere else.

Reply:
I would agree with you if the anonymous function were small and not recursive. I believe that moving this to a named function will improve the code in two different ways:
- It will make it easier to read. The function context will be concise and specific because will not be possible to read variables from the upper scope.
- The Go compiler will be able to optimize these recursive calls.

Reply:
Did you mean?
```go
// sortPkgs sorts the given packages by their dependencies.
func SortPkgs(pkgs []pkg) error {
	visited := make(map[string]bool)
	onStack := make(map[string]bool)
	sortedPkgs := make([]pkg, 0, len(pkgs))

	// Visit all packages
	for _, p := range pkgs {
		if err := visitPackage(p, pkgs, visited, onStack, &sortedPkgs); err != nil {
			return err
		}
	}

	copy(pkgs, sortedPkgs)
	return nil
}

// visitNode visits a package's and its dependencies dependencies and adds them to the sorted list.
func visitPackage(pkg pkg, pkgs []pkg, visited, onStack map[string]bool, sortedPkgs *[]pkg) error {
	if onStack[pkg.name] {
		return fmt.Errorf("cycle detected: %s", pkg.name)
	}
	if visited[pkg.name] {
		return nil
	}

	visited[pkg.name] = true
	onStack[pkg.name] = true

	// Visit package's dependencies
	for _, req := range pkg.requires {
		found := false
		for _, p := range pkgs {
			if p.name != req {
				continue
			}
			if err := visitPackage(p, pkgs, visited, onStack, sortedPkgs); err != nil {
				return err
			}
			found = true
			break
		}
		if !found {
			return fmt.Errorf("missing dependency '%s' for package '%s'", req, pkg.name)
		}
	}

	onStack[pkg.name] = false
	*sortedPkgs = append(*sortedPkgs, pkg)
	return nil
}
```

Reply:
Resolved-by: 312236f1114c8b606ed724dcb0926d1444feeba1